### PR TITLE
feat: Add restic check support

### PR DESCRIPTION
**Pull request by Kaitou786**
Closes: #123
Signed-off-by: Tarun Khandelwal <tarun@flanksource.com>

### CODE REVIEW

File: checks/restic.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+func (c *ResticChecker) Check(extConfig external.Check) *pkg.CheckResult {
+	logger.Debugf("Starting Restic checks")
+	start := time.Now()
+	resticCheck := extConfig.(v1.ResticCheck)
+	envVars := getEnvVars(resticCheck)
+	if err := checkIntegrity(resticCheck.Repository, resticCheck.CaCert, envVars); err != nil {
+		return Failf(resticCheck, "failed to check integrity of repository %v", err)
+	}
```
Comment (moshloop) [inline checks/restic.go:46]: This should be optional

File: checks/restic.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+func (c *ResticChecker) Run(config v1.CanarySpec) []*pkg.CheckResult {
+	var results []*pkg.CheckResult
+	for _, conf := range config.Restic {
+		results = append(results, c.Check(conf))
+	}
+	return results
+}
+
+func (c *ResticChecker) Check(extConfig external.Check) *pkg.CheckResult {
+	logger.Debugf("Starting Restic checks")
```
Comment (moshloop) [inline checks/restic.go:40]:
```suggestion
```

File: checks/restic.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	resticCheck := extConfig.(v1.ResticCheck)
+	envVars := getEnvVars(resticCheck)
+	if err := checkIntegrity(resticCheck.Repository, resticCheck.CaCert, envVars); err != nil {
+		return Failf(resticCheck, "failed to check integrity of repository %v", err)
+	}
+	logger.Debugf("Restic integrity check succeeded for repository: %s", resticCheck.Repository)
+	if err := checkBackupFreshness(resticCheck.Repository, resticCheck.MaxAge, resticCheck.CaCert, envVars); err != nil {
+		return Failf(resticCheck, "failed while checking backup freshness: %v", err)
+	}
+	logger.Debugf("Restic backup freshness check succeeded for repository: %s", resticCheck.Repository)
```
Comment (moshloop) [inline checks/restic.go:51]: This type of logging should be handled in Success or in the outer function
```suggestion
```

File: checks/restic.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	logger.Debugf("Starting Restic checks")
+	start := time.Now()
+	resticCheck := extConfig.(v1.ResticCheck)
+	envVars := getEnvVars(resticCheck)
+	if err := checkIntegrity(resticCheck.Repository, resticCheck.CaCert, envVars); err != nil {
+		return Failf(resticCheck, "failed to check integrity of repository %v", err)
+	}
+	logger.Debugf("Restic integrity check succeeded for repository: %s", resticCheck.Repository)
+	if err := checkBackupFreshness(resticCheck.Repository, resticCheck.MaxAge, resticCheck.CaCert, envVars); err != nil {
+		return Failf(resticCheck, "failed while checking backup freshness: %v", err)
```
Comment (moshloop) [inline checks/restic.go:49]: This shouldn't return an error, but rather a message with how much older it is than than the maxage

File: checks/restic.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	return results
+}
+
+func (c *ResticChecker) Check(extConfig external.Check) *pkg.CheckResult {
+	logger.Debugf("Starting Restic checks")
+	start := time.Now()
+	resticCheck := extConfig.(v1.ResticCheck)
+	envVars := getEnvVars(resticCheck)
+	if err := checkIntegrity(resticCheck.Repository, resticCheck.CaCert, envVars); err != nil {
+		return Failf(resticCheck, "failed to check integrity of repository %v", err)
```
Comment (moshloop) [inline checks/restic.go:45]:
```suggestion
		return Failf(resticCheck, "integrity check failed: %v", err)
```

File: api/v1/checks.go [hunk 6]
```diff
@@ -163,6 +163,35 @@ func (c S3BucketCheck) GetType() string {
...
+	// Password for the restic repository
+	Password string `yaml:"password" json:"password"`
+	// MaxAge for backup freshness
+	MaxAge string `yaml:"maxAge" json:"maxAge"`
+	// AccessKey access key id for connection with aws s3, minio, wasabi, alibaba oss
+	AccessKey string `yaml:"accessKey,omitempty" json:"accessKey,omitempty"`
+	// SecretKey secret access key for connection with aws s3, minio, wasabi, alibaba oss
+	SecretKey string `yaml:"secretKey,omitempty" json:"secretKey,omitempty"`
+	// CaCert path to the root cert. In case of self-signed certificates
+	CaCert string `yaml:"caCert,omitempty" json:"caCert,omitempty"`
```
Comment (moshloop) [inline api/v1/checks.go:180]: This should probably be InsecureTLSSkipVerify like in http checlk

File: checks/restic.go (hunk 1 above)
Comment (Kaitou786) [reply checks/restic.go:46]: So like exposing a CRD attribute `checkIntegrity` if set to true will check this?

File: checks/restic.go (hunk 4 above)
Comment (Kaitou786) [reply checks/restic.go:49]: Okay... I can include how much the backup is older than the maxAge but the canary should fail in that case, right?
So Only the  return message needs to be adjusted right?
The `Failf` remains here, if I got it correctly?

File: api/v1/checks.go (hunk 6 above)
Comment (Kaitou786) [reply api/v1/checks.go:180]: Okay... though in restic skipping a [TLS/Verification is not possible](https://github.com/restic/restic/issues/2656) user needs to define the path of CaRoot cert. I took this field as same in helmCheck

File: checks/restic.go (hunk 4 above)
Comment (moshloop) [reply checks/restic.go:49]: Correct

File: api/v1/checks.go [hunk 7]
```diff
@@ -163,6 +163,37 @@ func (c S3BucketCheck) GetType() string {
...
+type ResticCheck struct {
+	// Description about the current restic canary check
+	Description string `yaml:"description,omitempty" json:"description,omitempty"`
+	// Repository The restic repository path eg: rest:https://user:pass@host:8000/ or rest:https://host:8000/ or s3:s3.amazonaws.com/bucket_name
+	Repository string `yaml:"repository" json:"repository"`
+	// Password for the restic repository
+	Password string `yaml:"password" json:"password"`
+	// MaxAge for backup freshness
+	MaxAge string `yaml:"maxAge" json:"maxAge"`
+	// CheckIntegrity when enabled will check the Integrity and consistency of the restic reposiotry
```
Comment (moshloop) [inline api/v1/checks.go:175]: Is this repository or the latest snapshot ?

Comment (Kaitou786) [reply api/v1/checks.go:175]: The MaxAge is for the latest snapshot.... the integrity check is for the whole repository

Comment (moshloop) [reply api/v1/checks.go:175]: We don't want to run an integrity check on the whole repo, as it can easily grow to TB's - Only on the most recent snapshot

**Pull request merged**